src/bindings/python/flux/kvs.py

```diff
@@ -0,0 +1,216 @@
+from _kvs import ffi, lib
+import flux
+from flux.wrapper import Wrapper, WrapperPimpl, WrapperBase
+import flux.json_c as json_c
+import json
+import collections
+import contextlib
+import errno
+
+raw = Wrapper(ffi, lib, prefixes=['kvs','kvs_'])
+ 
+
+class KVSHandle(Wrapper):
+  def __init__(self, flux_handle=None):
+    if flux_handle is None:
+      raise ValueError("flux_handle must be a valid Flux object")
+    super(self.__class__, self).__init__(ffi, lib, handle=flux_handle,
+                                     match=ffi.typeof(lib.flux_open).result,
+                                     prefixes=[
+                                       'kvs_',
+                                       ],
+                                     )
+
+class KVSDir(WrapperPimpl,collections.MutableMapping):
+  class InnerWrapper(Wrapper):
+    def __init__(self, flux_handle=None, path='.', handle=None):
+      self.handle = None
+      global raw
+      if flux_handle is None and handle is None:
+        raise ValueError("flux_handle must be a valid Flux object or handle must be a valid kvsdir cdata pointer")
+      if isinstance(flux_handle, WrapperBase):
+        flux_handle = flux_handle.handle
+      if handle is None:
+        d = ffi.new("kvsdir_t [1]")
+        raw.kvs_get_dir(flux_handle, d, path)
+        handle = d[0]
+        if handle is None or handle == ffi.NULL:
+          raise IOError(errno.ENOENT, os.strerror(errno.ENOENT))
+
+      super(self.__class__, self).__init__(ffi, lib, handle=handle,
+                                       match=ffi.typeof('kvsdir_t'),
+                                       prefixes=[
+                                         'kvsdir_',
+                                         ],
+                                       )
+    def __del__(self):
+      if self.handle is not None:
+        raw.kvsdir_destroy(self.handle)
+
+  def __init__(self, flux_handle=None, path='.', handle=None):
+      self.fh = flux_handle
+      self.path = path
+      if flux_handle is None and handle is None:
+        raise ValueError("flux_handle must be a valid Flux object or handle must be a valid kvsdir cdata pointer")
+      self.pimpl = self.InnerWrapper(flux_handle, path, handle)
+
+  def commit(self):
+    global raw
+    ret = raw.kvs_commit(self.fh.handle)
+    if ret < 0:
+      raise EnvironmentError(ret)
+
+  def key_at(self, key):
+    c_str = self.pimpl.key_at(key)
+    p_str = ffi.string(c_str)
+    lib.free(c_str)
+    return p_str
+
+  def __getitem__(self, key):
+    try:
+      j = json_c.Jobj()
+      self.pimpl.get(key, j.get_as_dptr())
+      return json.loads(j.as_str())
+    except EnvironmentError as err:
+      if err == errno.EISDIR:
```

Comment:
This one surprises me a bit.  You're absolutely right about the 
err.errno issue, I'll push up that fix in just a moment.  The surprise 
is that running your example this is what I get after doing a `flux 
wreckrun -N 4 hostname`:

```
         cmdline [u'hostname']
         nnodes 4
         cwd /g/g12/scogland/projects/flux/flux-core
         ntasks 1
         environ {u'_ZSH_TMUX_FIXED_CONFIG': 
u'/g/g12/scogland/.zsh/oh-my-zsh/plugins/tmux/tmux.extra.conf', u'LESS': 
u'-R', u'INFOPATH': u'/usr/apps/gnu/4.8.2/share/info', u'SHELL': 
u'/bin/zsh', u'HISTSIZE': u'10000', u'VIRTUAL_ENV_DISABLE_PROMPT': u'1', 
u'DATE_FMT': u'%c', u'MANPATH': 
u'/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/share/man:/usr/apps/gnu/4.8.2/share/man:/usr/local/tools/dotkit/dotkit/man:/usr/man:/usr/share/man:/usr/local/man:/usr/X11R6/man:/usr/lib64/mvapich/default/man:/collab/usr/global/tools/git/chaos_5_x86_64_ib/git-2.0.0/share/man', 
u'TASKDDATA': u'/g/g12/scogland/.task_server_data', u'PYTHONPATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/lib/python2.7/site-packages', 
u'FPATH': 
u'/g/g12/scogland/.zsh/mine/funcs:/g/g12/scogland/.zsh/funcs:/g/g12/scogland/.zsh/oh-my-zsh/plugins/z:/g/g12/scogland/.zsh/oh-my-zsh/plugins/tmux:/g/g12/scogland/.zsh/oh_my_zsh_custom/plugins/hub:/g/g12/scogland/.zsh/oh_my_zsh_custom/plugins/taskwarrior:/g/g12/scogland/.zsh/oh-my-zsh/plugins/screen:/g/g12/scogland/.zsh/oh-my-zsh/plugins/osx:/g/g12/scogland/.zsh/oh-my-zsh/plugins/brew:/g/g12/scogland/.zsh/oh-my-zsh/plugins/history-substring-search:/g/g12/scogland/.zsh/oh-my-zsh/plugins/git-remote-branch:/g/g12/scogland/.zsh/oh-my-zsh/plugins/git-extras:/g/g12/scogland/.zsh/oh-my-zsh/plugins/git:/g/g12/scogland/.zsh/oh-my-zsh/plugins/svn-fast-info:/g/g12/scogland/.zsh/oh-my-zsh/plugins/gitfast:/g/g12/scogland/.zsh/oh-my-zsh/functions:/g/g12/scogland/.zsh/oh-my-zsh/completions:/usr/local/tools/dotkit/dotkit/ksh:/usr/share/zsh/site-functions:/usr/share/zsh/4.3.10/functions', 
u'GITHUB_USER': u'trws', u'HOMEBREW_GITHUB_API_TOKEN': 
u'e68741bb043902338ab33f78f8f88da573f66e6a', u'SSH_TTY': u'/dev/pts/12', 
u'MAIL': u'/var/spool/mail/scogland', u'SSH_CONNECTION': 
u'10.247.249.134 59678 134.9.38.5 22', u'HOST_GRP': u'linux', 
u'SSH_ASKPASS': u'/usr/libexec/openssh/gnome-ssh-askpass', u'GOPATH': 
u'/g/g12/scogland/go', u'FIREFOX_OPEN_IN': u'tab', u'GOROOT': 
u'/g/g12/scogland/go', u'FLUX_CONNECTOR_PATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/lib/flux/connectors', 
u'LESSOPEN': u'|lesspipe.sh %s', u'CVS_RSH': u'ssh', u'USER': 
u'scogland', u'INPUTRC': u'/etc/inputrc', u'QTDIR': 
u'/usr/lib64/qt-3.3', u'ZSH_TMUX_TERM': u'screen-256color', u'SHLVL': 
u'3', u'DK_NODE': 
u'/g/g12/scogland/spack/share/spack/dotkit/chaos_5_x86_64_ib:/usr/local/dpkg-db/info:/usr/local/etc/dotkits:/usr/global/tools/dotkit', 
u'MPIRUN_RSH_LAUNCH': 1, u'DK_UEQRU': u'1', u'EDITOR': u'vim', 
u'MODULESHOME': u'/usr/share/Modules', u'EPREFIX': 
u'/g/g12/scogland/programs', u'QTLIB': u'/usr/lib64/qt-3.3/lib', 
u'LUA_CPATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/lib64/lua/5.1/?.so;;;', 
u'TMUX': u'/tmp/scogland/tmux-37084/default,141890,0', u'ENV': 
u'/g/g12/scogland/.kshrc', u'TMPDIR': u'/var/tmp/scogland', 
u'MODULEPATH': 
u'/g/g12/scogland/spack/share/spack/modules/chaos_5_x86_64_ib:/usr/share/Modules/modulefiles:/etc/modulefiles:/opt/modules/modulefiles', 
u'DK_SUBNODE': u'ksh/chaos_5_x86_64_ib:ksh:chaos_5_x86_64_ib:.', 
u'LS_COLORS': 
u'no=00:fi=00:di=01;34:ln=01;36:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.gz=01;31:*.bz2=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.avi=01;35:*.fli=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.ogg=01;35:*.mp3=01;35:*.wav=01;35:', 
u'PAGER': u'less', u'TMP': u'/var/tmp/scogland', u'DUALCASE': u'1', 
u'OMP_NESTED': u'true', u'_dk_shell': u'ksh', u'HOME': 
u'/g/g12/scogland', u'LD_LIBRARY_PATH': 
u'/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/lib:/g/g12/scogland/spack/opt/spack/chaos_5_x86_64_ib/gcc-4.9.2/python-2.7.10-drsaoxceqqgx2yoovbzkp3hnau3si5bg/lib', 
u'LANG': u'en_US.UTF-8', u'PFTP_CONFIG_FILENAME': u'/etc/pftp_config', 
u'FLUX_MODULE_PATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/lib/flux/modules', 
u'LIBRARY_PATH': 
u'/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/lib', 
u'_dk_rl': u'-1', u'QTINC': u'/usr/lib64/qt-3.3/include', 
u'GIT_AUTHOR_EMAIL': u'tscogland@llnl.gov', u'G_BROKEN_FILENAMES': u'1', 
u'LUA_PATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/share/lua/5.1/?.lua;;;', 
u'NNTPSERVER': u'snews://news.purdue.edu', u'NSS_STRICT_NOFORK': 
u'DISABLED', u'BASH_FUNC_module()': u'() {  eval `/usr/bin/modulecmd 
bash $*`\n}', u'_dk_isatty': u'1', u'KDEDIRS': u'/usr', 
u'vcs_info_msg_0_': 
u'(%{%F{81}%}python_mod%{%F{161}%}\u25cf%{\x1b[00m%}) ', 
u'PERLBREW_HOME': u'/g/g12/scogland/.perlbrew', u'ENTOMB': u'yes', 
u'TMUX_PANE': u'%12', u'SYSTEM': u'Linux', u'SSH_CLIENT': 
u'10.247.249.134 56887 22', u'PERLBREW_BASHRC_VERSION': u'0.73', 
u'LOGNAME': u'scogland', u'PATH': 
u'/usr/global/tools/clang/chaos_5_x86_64_ib/clang-omp-3.5.0/bin:/usr/apps/gnu/4.8.2/bin:/usr/local/tools/ld-auto-rpath/bin:/g/g12/scogland/perl5/perlbrew/bin:/g/g12/scogland/scripts:/g/g12/scogland/programs/bin:/g/g12/scogland/spack/bin:/g/g12/scogland/go/bin:~/programs/bin:/usr/lib64/qt-3.3/bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/sbin:.:/g/g12/scogland/bin:/usr/global/tools/totalview/m/hype/default/bin:/collab/usr/global/tools/git/chaos_5_x86_64_ib/git-2.0.0/bin', 
u'BIN_SH': u'xpg4', u'HISTCONTROL': u'ignoredups', u'TERM': 
u'xterm-256color', u'FLUX_BROKER_PATH': 
u'/g/g12/scogland/projects/flux/flux-core/install/libexec/flux/flux-broker', 
u'FLUX_SEC_DIRECTORY': 
u'/g/g12/scogland/projects/flux/flux-core/src/cmd/.libs/../../../etc/flux', 
u'OMP_MAX_ACTIVE_LEVELS': u'5', u'FLUX_CONF_DIRECTORY': 
u'/g/g12/scogland/projects/flux/flux-core/src/cmd/.libs/../../../etc/flux', 
u'TOTALVIEW': u'totalview', u'PERLBREW_ROOT': 
u'/g/g12/scogland/perl5/perlbrew', u'SSH_AUTH_SOCK': 
u'/tmp/ssh-vteW122225/agent.122225', u'KRB5CCNAME': 
u'FILE:/tmp/krb5cc_37084_tabgqx', u'GIT_AUTHOR_NAME': u'Tom Scogland', 
u'LC_ALL': u'en_US.UTF-8', u'WISECONFIGDIR': u'/usr/share/wise2/', 
u'SYS_TYPE': u'chaos_5_x86_64_ib', u'_dk_inuse': u'clang-omp-3.5.0.0 
gcc-4.8.2p.1 ld-auto-rpath.2 lcinit.0 reuse.1 tv.1 git.1', u'LSCOLORS': 
u'Gxfxcxdxbxegedabagacad', u'LOADEDMODULES': u'', u'_Z_DATA': 
u'/g/g12/scogland/.zsh/z-dirjump-list.txt', u'GLOGIN': u'1', u'PWD': 
u'/g/g12/scogland/projects/flux/flux-core', u'DK_ROOT': 
u'/usr/local/tools/dotkit/dotkit'}
         create-time 2015-07-08T13:32:36
         rank <flux.kvs.KVSDir object at 0x2aaab5e331d0>
         starting-time 2015-07-08T13:32:36
         running-time 2015-07-08T13:32:36
         state complete
         complete-time 2015-07-08T13:32:36
         0 <flux.kvs.KVSDir object at 0x2aaab5e33150>
```

Perhaps it should be set up to print the dirs as dicts, but for now the 
objects are printing as they have been programmed to.  The only issue 
should be if the value of the thing cannot be read with `kvs_get` and 
decoded from the resulting json object, but I haven't run into anything 
yet that has that issue.  Are you running the most recent version?

I just pushed up a new feature that's analogous to the `os.walk()` 
routine, as in it is a generator that walks the entire directory tree 
under a path.  For example, this script prints the entire KVS space's 
keys and values from the top down. The one thing it doesn't do is handle 
symlinks, because I'm not sure how best to resolve those just now.

```python
import flux
from flux.core import Flux
import flux.kvs as kvs

h = Flux()
with kvs.get_dir(h, ".") as kd:
   for root, dirs, files in kvs.walk(kd, topdown=True):
     print root
     for f in files:
       print root+f, '=', kd[root+f]
```

On 8 Jul 2015, at 13:21, Stephen Herbein wrote:

>> +    if ret < 0:
>> +      raise EnvironmentError(ret)
>> +
>> +  def key_at(self, key):
>> +    c_str = self.pimpl.key_at(key)
>> +    p_str = ffi.string(c_str)
>> +    lib.free(c_str)
>> +    return p_str
>> +
>> +  def __getitem__(self, key):
>> +    try:
>> +      j = json_c.Jobj()
>> +      self.pimpl.get(key, j.get_as_dptr())
>> +      return json.loads(j.as_str())
>> +    except EnvironmentError as err:
>> +      if err == errno.EISDIR:
>
> So this doesn't work as expected for kvs directories, but the below 
> should fix it:
> ```python
> if err.errno == errno.EISDIR:
> ```
>
> Unfortunately, I'm still getting Key/Env Errors on 
> non-integer/non-directory values. I tried iterating over the values in 
> the lwj directory (lwj.1.nnodes, lwj.1.create-time, etc) and I get 
> errors for any non-integer value.  Most give me KeyErrors, but one 
> (and sub-directory) gives me an EnvError which passes through b/c the 
> errno != EISDIR. The minimal test case along with the output is below:
> ```python
>  with kvs.get_dir(h, "lwj.1") as lwj_dir:
>      for prop_name in lwj_dir:
>              try:
>                  print "\t{} {}".format(prop_name, lwj_dir[prop_name])
>              except EnvironmentError as err:
>                  print "EnvError: {}, Key: {}".format(err, prop_name)
>              except KeyError as err:
>                  print "KeyError:", err
> ```
> Note: All of the errors below have a errno of 9: Bad file descriptor
> ```
>      nnodes 8
>      ntasks 128
> KeyError: 'create-time'
> KeyError: 'user'
> KeyError: 'jobname'
> KeyError: 'account'
> KeyError: 'submit_time'
> KeyError: 'execution_time'
> KeyError: 'time_limit'
>      ncpus 128
>      io_rate 2
> KeyError: 'allocated-time'
> EnvError: [Errno 9] Bad file descriptor, Key: rank
> KeyError: 'runrequest-time'
> KeyError: 'starting_time'
> KeyError: 'running_time'
> KeyError: 'state'
> KeyError: 'complete_time'
> KeyError: 'io_time'
> ```
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/236/files#r34193606